### PR TITLE
[Proposal] Image segmentation pipeline

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?
- Currently very low-level results, simply classification masks + score
+ label for each detected class
- Could support panoptic, instance segmentation, bounding boxes and even
part of instance segmentation (might require adding the "parent" info in
addition but that's about it.)
- Happy to hear some thoughts about the design.
- Future could maybe add a "aggregation_strategy" like, which could maybe
output a single image with color filters on top of the classes and so on
(or leave this reduction outside the pipeline). Getting final images like
https://colab.research.google.com/drive/16jcaJoc6bCFAQ96jDe2HwtXj7BMD_-m5#scrollTo=8IRGo8d0qkgR 
is still a bit involved for users that simply want to "see" outputs.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@LysandreJik @NielsRogge

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/image_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        if not is_batched:
+            images = [images]
+
+        images = [self.load_image(image) for image in images]
+
+        with torch.no_grad():
+            inputs = self.feature_extractor(images=images, return_tensors="pt")
+            outputs = self.model(**inputs)
+
+            processed_sizes = [(im.height, im.width) for im in images]
```
Comment (NielsRogge) [inline src/transformers/pipelines/image_segmentation.py:123]: In case images are PIL images, then width comes first, so it should be (im.width, im.height).

For (PyTorch) tensors, height comes first.

**Comment by LysandreJik** [comment]
@NielsRogge can you give this pipeline a look when you get a chance?

**Comment by NielsRogge** [comment]
Thanks for this proposal! So the output of the API would be `[{mask, score, label}, ...]`, right? What is the shape of each of those?

I wonder whether this can support any image segmentation model. I'm currently working on implementing [SegFormer](https://arxiv.org/abs/2105.15203) (a new model by NVIDIA), which is a semantic segmentation model. It takes an image as input and produces a segmentation map as output (i.e. it assigns a class to each pixel). This model will only have logits, and they are of shape (batch_size, num_labels, height/4, width/4) - the logits are produced at 1/4th of the original image size. The SegFormer model does not have `pred_masks`, for example. 

This pipeline should support all forms of image segmentation, right? Panoptic, semantic, instance, etc?

**Comment by Narsil** [comment]
@NielsRogge 

I think pipelines should support as much different models as possible, regardless of the model output. (and clean error when model is not supported)
Proposed implem uses `pred_masks` because that's what currently available for DetR but if some other arch have different outputs, it should be the pipeline's role to be able to use `.logits` instead and still produce the same outputs.

Those are different kind of models, right ? (Like XXXXForSegmentation vs XXXXForPanopticSegmentation)
If yes, then I think it's kind of routine to have switches on those for the pipelines. Less desirable would be to switch on actual model outputs (.logits vs .pred_masks) , and something that we should really strive to avoid is actually looking at the model arch to decide.

Regardless, I think we can recover masks from raw logits, right ? If yes, then I think that proves that current output is good as it would enable pipeline to support SegFormer too.

Would you agree?

**Comment by NielsRogge** [comment]
Unstale
